azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +397,139 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
-        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
-        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        Returns True if the provided controller is enabled at the provided cgroup.
+
+        There are two ways to determine if a controller is enabled at the provided cgroup:
+
+        1. For non-leaf cgroups, the cgroup.subtree_control shows space separated list of the controllers which are
+        enabled to control resource distribution from the cgroup to its children. All non-root "cgroup.subtree_control"
+        files can only contain controllers which are enabled in the parent's "cgroup.subtree_control" file.
+                $ cat /sys/fs/cgroup/cgroup.subtree_control
+                cpuset cpu io memory hugetlb pids rdma misc
+
+        2. For leaf cgroups, the cgroup.subtree_control file will be empty and the presence of "<controller>."
+        prefixed interface files at the path indicate the controller is enabled.
+                $ ls /sys/fs/cgroup/azure.slice/walinuxagent.service/
+                cgroup.controllers  cgroup.max.descendants  cgroup.threads  cpu.pressure    cpu.weight.nice      memory.high       memory.oom.group  memory.swap.current  memory.zswap.current  pids.peak
+                cgroup.events       cgroup.pressure         cgroup.type     cpu.stat        io.pressure          memory.low        memory.peak       memory.swap.events   memory.zswap.max
+                cgroup.freeze       cgroup.procs            cpu.idle        cpu.uclamp.max  memory.current       memory.max        memory.pressure   memory.swap.high     pids.current
+                cgroup.kill         cgroup.stat             cpu.max         cpu.uclamp.min  memory.events        memory.min        memory.reclaim    memory.swap.max      pids.events
+                cgroup.max.depth    cgroup.subtree_control  cpu.max.burst   cpu.weight      memory.events.local  memory.numa_stat  memory.stat       memory.swap.peak     pids.max
+
+        If either check is True, the controller is enabled at the cgroup. Check 1 is necessary because no controller
+        interface files exist at the root cgroup, even if the controller is enabled.
         """
-        return CGroupsApi._foreach_legacy_cgroup(lambda *_: None)
+        if cgroup_path is not None and controller is not None:
+            # Check that the controller is enabled in the cgroup.subtree_control file
+            enabled_controllers_file = os.path.join(cgroup_path, 'cgroup.subtree_control')
+            if os.path.exists(enabled_controllers_file):
+                enabled_controllers = fileutil.read_file(enabled_controllers_file).rstrip().split(" ")
+                if controller in enabled_controllers:
+                    return True
+
+            # Check that the controller interface files exist in the cgroup
+            if os.path.exists(cgroup_path):
```

Comment:
cpu.stat is a special case. From controllers documentation, it's the only controller interface file which exists whether controller enabled or not.

Also CpuAccounting is a bit different in v2. From systemd documentation:
> 
> CPUAccounting=
> Turn on CPU usage accounting for this unit. Takes a boolean argument. Note that turning on CPU accounting for one unit will also implicitly turn it on for all units contained in the same slice and for all its parent slices and the units contained therein. The system default for this setting may be controlled with DefaultCPUAccounting= in [systemd-system.conf(5)](https://man.archlinux.org/man/systemd-system.conf.5.en).
> **Under the unified cgroup hierarchy, CPU accounting is available for all units and this setting has no effect.**

When we specify CPUAccounting=yes in slice/drop-in files in v2, the cpu controller isn't enabled. CPU controller will only be enabled in our use case once CPUQuota is set. cpu usage is available for all units, so I figured it was safe that we consider it enabled here since cpu.stat is available. But I understand it may lead to confusion

If this is removed, cgroupconfigurator would need to have some other logic to determine if cpu controller is available and cpu quota could be set.